pkg/app/piped/imagewatcher/watcher.go

```diff
@@ -63,108 +65,113 @@ func NewWatcher(cfg *config.PipedSpec, gitClient gitClient, logger *zap.Logger)
 	}
 }
 
-// Run spawns goroutines for each image provider. They periodically pull the image
+// Run spawns goroutines for each git repository. They periodically pull the image
 // from the container registry to compare the image with one in the git repository.
 func (w *watcher) Run(ctx context.Context) error {
-	// TODO: Spawn goroutines for each repository
-	// Pre-clone to cache the registered git repositories.
-	for _, r := range w.config.Repositories {
-		// TODO: Clone repository another temporary destination
-		repo, err := w.gitClient.Clone(ctx, r.RepoID, r.Remote, r.Branch, "")
-		if err != nil {
-			w.logger.Error("failed to clone repository",
-				zap.String("repo-id", r.RepoID),
-				zap.Error(err),
-			)
-			return fmt.Errorf("failed to clone repository %s: %w", r.RepoID, err)
-		}
-		w.gitRepos[r.RepoID] = repo
-	}
-
+	w.providerCfgs = make(map[string]config.PipedImageProvider, len(w.config.ImageProviders))
 	for _, cfg := range w.config.ImageProviders {
-		p, err := imageprovider.NewProvider(&cfg, w.logger)
-		if err != nil {
-			return fmt.Errorf("failed to yield image provider %s: %w", cfg.Name, err)
-		}
+		w.providerCfgs[cfg.Name] = cfg
+	}
 
+	for _, repo := range w.config.Repositories {
 		w.wg.Add(1)
-		go w.run(ctx, p, cfg.PullInterval.Duration())
+		go w.run(ctx, &repo)
 	}
+
 	w.wg.Wait()
 	return nil
 }
 
-// run periodically compares the image in the given provider and one in git repository.
+// run periodically compares the image in the given git repository and one in the image provider.
 // And then pushes those with differences.
-func (w *watcher) run(ctx context.Context, provider imageprovider.Provider, interval time.Duration) {
+func (w *watcher) run(ctx context.Context, repoCfg *config.PipedRepository) {
 	defer w.wg.Done()
-	ticker := time.NewTicker(interval)
-	defer ticker.Stop()
 
+	var (
+		pullInterval = defaultPullInterval
+		commitMsg    = ""
+		includedCfgs = []string{}
+		excludedCfgs = []string{}
+	)
+	// Use user-defined settings if there is.
+	for _, r := range w.config.ImageWatcher.Repos {
+		if r.RepoID != repoCfg.RepoID {
+			continue
+		}
+		pullInterval = time.Duration(r.PullInterval)
+		commitMsg = r.CommitMessage
+		includedCfgs = r.Includes
+		excludedCfgs = r.Excludes
+		break
+	}
+
+	// Periodically update this cloned directory as long as this worker continues.
+	repo, err := w.gitClient.Clone(ctx, repoCfg.RepoID, repoCfg.Remote, repoCfg.Branch, "")
+	if err != nil {
```

Comment:
When one of the goroutine stops due to this error, should we cancel all of the others?
Currently, this one is stopped and the other ones still running and just an error was logged.

Reply:
Right, we should ensure all workers properly start running. Using sync/errgroup would be better.